### DISCUSSION TITLE
Entire Workflow is validated instead of delta change

### DISCUSSION THREAD

**Discussion by vaibhavdangar-woolies** [scolladon/sfdx-git-delta#358] [Q&A] [answered]
Ref: https://github.com/scolladon/sfdx-git-delta/issues/38

Hi @scolladon ,

Query: Even for a 'WorkflowAlert' destructive change is it necessary for the plugin to show 'Workflow' name in package.xml?
Since the workflow name is listed in package.xml it performs entire file validation, rather than delta (destructive change in this case) only.
Please advise.

**Comment by vaibhavdangar-woolies** [comment]


**Reply by vaibhavdangar-woolies** [reply]
Is it possible to not have the workflow included in package.xml

**Reply by vaibhavdangar-woolies** [reply]
![image](https://user-images.githubusercontent.com/48708874/192720112-44829695-c0b9-4bd9-80e8-f718e294a9c6.png)

**Reply by scolladon** [reply]
Hi @vaibhavdangar-woolies 

could you provide the result of those commands please :
```sh
$ git diff --name-status --no-renames <from> <to>
$ git diff --no-prefix <from> <to> -- force-app/main/default/workflows/ABC.workflow-metadata.xml 
```

By the way, the workflow file extension looks weird to me, shouldn't it be `workflow-meta.xml` ?

**Reply by vaibhavdangar-woolies** [reply]
It is force-app/main/default/workflows/ABC.workflow-meta.xml , typo while masking the filename.
Allow me to get you the git outputs.

**Reply by vaibhavdangar-woolies** [reply]
Hi @scolladon , please review outputs

```
**git diff --name-status --no-renames origin/feature-1-branch origin/base-branch**
OUTPUT:
M       force-app/main/default/workflows/ABC.workflow-meta.xml


**git diff --no-prefix origin/feature-1-branch origin/base-branch -- force-app/main/default/workflows/ABC.workflow-meta.xml**

OUTPUT:
diff --git force-app/main/default/workflows/ABC.workflow-meta.xml force-app/main/default/workflows/ABC.workflow-meta.xml
index e7884f9eba..6e0da18c39 100644
--- force-app/main/default/workflows/ABC.workflow-meta.xml
+++ force-app/main/default/workflows/ABC.workflow-meta.xml
@@ -11,27 +11,6 @@
         <senderType>DefaultWorkflowUser</senderType>
         <template>Case_Management_PQR/Additional_Info</template>
     </alerts>
-    <alerts>
-        <fullName>GC_1</fullName>
-        <ccEmails>USER@EMAIL.com.au</ccEmails>
-        <description>GC  1 </description>
-        <protected>false</protected>
-        <recipients>
-            <field>GC_Email__c</field>
-            <type>email</type>
-        </recipients>
-        <recipients>
-            <field>Category_Manager__c</field>
-            <type>userLookup</type>
-        </recipients>
-        <recipients>
-            <field>Merc__c</field>
-            <type>userLookup</type>
-        </recipients>
-        <senderAddress>test@hello.EMAIL.com.au</senderAddress>
-        <senderType>OrgWideEmailAddress</senderType>
-        <template>helloReminder_Template</template>  
-    </alerts>
     <alerts>
         <fullName>Add_Doc</fullName>
         <description>AddDoc</description>
@@ -284,6 +263,27 @@
         <senderType>OrgWideEmailAddress</senderType>
         <template>helloReminder_Templatee</template>
     </alerts>
+    <alerts>
+        <fullName>GC_1</fullName>
+        <ccEmails>USER@EMAIL.com.au</ccEmails>
+        <description>GC  1 </description>
+        <protected>false</protected>
+        <recipients>
+            <field>GC_Email__c</field>
+            <type>email</type>
+        </recipients>
+        <recipients>
+            <field>Category_Manager__c</field>
+            <type>userLookup</type>
+        </recipients>
+        <recipients>
+            <field>Merc__c</field>
+            <type>userLookup</type>
+        </recipients>
+        <senderAddress>test@hello.EMAIL.com.au</senderAddress>
+        <senderType>OrgWideEmailAddress</senderType>
+        <template>helloReminder_Template</template>  
+    </alerts>
     <alerts>
         <fullName>GC_1_day_after_the_Effective_date</fullName>
         <ccEmails>USER@EMAIL.com.au</ccEmails>
@@ -1596,20 +1596,6 @@
         <description>Email</description>
         <formula>masked</formula>
         <triggerType>onCreateOrTriggeringUpdate</triggerType>
-    </rules>
-       <rules>
-        <fullName>Update Case</fullName>
-        <actions>
-            <name>Update_Comments</name>
-            <type>FieldUpdate</type>
-        </actions>
-        <actions>
-            <name>Update_Status</name>
-            <type>FieldUpdate</type>
-        </actions>
-        <active>false</active>
-        <formula>masked</formula>
-        <triggerType>onCreateOrTriggeringUpdate</triggerType>
     </rules>
     <rules>
         <fullName>Populate Custom Case ID</fullName>
~




```

**Reply by vaibhavdangar-woolies** [reply]
Please find the corresponding package.xml file contents for above

```
####package.xml

<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
    <types>
        <members>ABC</members>
        <name>Workflow</name>
    </types>
    <types>
        <members>ABC.GC_1</members>
        <name>WorkflowAlert</name>
    </types>
    <types>
        <members>ABC.CaseClosed</members>
        <members>ABC.UpdateCase</members>
        <name>WorkflowRule</name>
    </types>
    <version>53.0</version>
</Package>


```

**Reply by vaibhavdangar-woolies** [reply]
Concern is why do we have the whole workflow file listed as change instead of individual components - workflow alert and workflow rule only being changed? 

```
<types>
        <members>ABC</members>
        <name>Workflow</name>
    </types>
```

The backdrop of this is we feel manual changes directly done in org. should not be overwritten, only the ones we pass as delta and seen in PR sould be deployed.

**Reply by scolladon** [reply]
Hi @vaibhavdangar-woolies 

Thanks for the details information.

The current behavior of the sfdx plugin is expected.
It is to cover an issue in the sfdx command we explained [here](https://github.com/scolladon/sfdx-git-delta/issues/99).

This issue is very old and maybe it is fixed.
If it is the case it means we no longer need to implement this workaround in the plugin and can remove adding the parent metadata in the package.xml

Could you test that it is possible to deploy only a `WorkflowRule` or a `WorkflowAlert` on your side (test if the previous sfdx issue is fixed) ? I'll do the same on my own.

**Reply by scolladon** [reply]
Something to know here as well

When using the sgd with those kind of files (workflow, sharing rule or custom labels) it will compute what has changed to create the package.xml. It will add the parent type in the package.xml as we have to workaround the sfdx deployability issue.

If you use the `-d` parameter it will also generate a source folder with only the changed elements and for workflow, sharing rule or custom label kinds of metadata, the file will only contain the changed elements. The plugin build those file with only what it needs to contain to deploy only what has changed. 

This way the Workflow type member in the package.xml will not have any incidence as it contains only what has changed anyway, and adding the workflow is just here to make sure the deployment is successful

It means for those files, deployment needs to be done from the generated incremental sources (using `force:source:deploy -p`) and not from the repository sources using the package.xml (using `force:source:deploy -x`) to benefit from the full trim of those files because of the sfdx deploy issue

**Comment by vaibhavdangar-woolies** [comment, marked as answer]
**Test results observtions:**

1.) Fails validation without the 'Workflow' node specification, just specifying the 'Workflowrule' or 'WorkflowAlert'

> 
> === Component Failures [3]
> 
>  Type  Name                                     Problem                                                                                                                                   
>  ───── ──────────────────────────────────────── ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────                
>  Error Case.GC_1     An object 'ABC.GC_1' of type WorkflowAlert was named in package.xml, but was not found in zipped directory
>  Error Case.UpdateCase An object 'ABC.UpdateCase' of type WorkflowRule was named in package.xml, but was not found in zipped directory
>  Error Case.CaseClosed              An object 'ABC.CaseClosed' of type WorkflowRule was named in package.xml, but was not found in zipped directory](url)
> 
> 


Need to include 'Workflow' as well
`    <types>
        <members>ABC</members>
        <name>Workflow</name>
    </types>
`

2.) On deployment, with 'Workflow' , 'WorkflowAlert', 'WorkflowRule' modifiications are only done to objects named in package.xml and does not alter anything else.

> *** Deploying with SOAP API v53.0 ***
DEPLOY PROGRESS | ████████████████████████████████████████ | 4/4 Components

=== Deployed Source

 FULL NAME                                TYPE          PROJECT PATH                                            
 ──────────────────────────────────────── ───────────── ─────────────────────────────────────────────────────── 
 ABC                                     Workflow      force-app/main/default/workflows/ABC.workflow-meta.xml
 ABC.GC_1     WorkflowAlert force-app/main/default/workflows/ABC.workflow-meta.xml
 ABC.CaseClosed              WorkflowRule  force-app/main/default/workflows/ABC.workflow-meta.xml
 ABC.UpdateCase WorkflowRule  force-app/main/default/workflows/ABC.workflow-meta.xml
Deploy Succeeded.

**Reply by scolladon** [reply]
That is what I have as well on my side
My version of sfdx is `sfdx-cli/7.168.0 darwin-x64 node-v14.19.3`

Still not fixed yet, we need to keep this behaviour then.
Using `-d` is a good way to deal with that and have only what is changed as explained [here](https://github.com/scolladon/sfdx-git-delta/discussions/358#discussioncomment-3754952)

**Reply by vaibhavdangar-woolies** [reply]
Thanks you for your quick response and assistance ! I'll close this query.
